test/JoeBarV2.test.ts

```diff
@@ -0,0 +1,147 @@
+import { ethers, network } from "hardhat"
+import { expect } from "chai"
+import { duration, increase } from "./utilities"
+import { describe } from "mocha"
+const hre = require("hardhat")
+
+describe("JoeBar", function () {
+  before(async function () {
+    this.JoeToken = await ethers.getContractFactory("JoeToken")
+    this.JoeBarV2 = await ethers.getContractFactory("JoeBarV2")
+    this.BarRewarder = await ethers.getContractFactory("BarRewarderPerSec")
+
+    this.signers = await ethers.getSigners()
+    this.dev = this.signers[0]
+    this.alice = this.signers[1]
+    this.bob = this.signers[2]
+    this.carol = this.signers[3]
+  })
+
+  beforeEach(async function () {
+    this.joe = await this.JoeToken.deploy()
+
+    this.rewarder = await this.BarRewarder.deploy(this.joe.address, "1500")
+
+    this.bar = await hre.upgrades.deployProxy(this.JoeBarV2, [this.joe.address, this.rewarder.address, "500"])
+
+    await this.bar.setRewarder(this.rewarder.address)
+    await this.joe.mint(this.alice.address, "100000000000000000000")
```

Comment:
nit: suggestion to use `ethers.utils.parseEther("100")` to make amount easier to read. similar for rest of file